vxquery-core/src/main/java/org/apache/vxquery/jsonparser/JSONParser.java

```diff
@@ -0,0 +1,188 @@
+import org.apache.htrace.fasterxml.jackson.core.JsonToken;
+import org.apache.hyracks.data.std.primitive.UTF8StringPointable;
+import org.apache.hyracks.data.std.util.ArrayBackedValueStorage;
+import org.apache.vxquery.datamodel.builders.atomic.StringValueBuilder;
+import org.apache.vxquery.datamodel.builders.jsonitem.ArrayBuilder;
+import org.apache.vxquery.datamodel.builders.jsonitem.ObjectBuilder;
+import org.apache.vxquery.datamodel.values.ValueTag;
+
+public class JSONParser {
+
+    final JsonFactory factory;
+    protected final ArrayBackedValueStorage atomic;
+    protected final List<ArrayBuilder> abStack;
+    protected final List<ObjectBuilder> obStack;
+    protected final List<ArrayBackedValueStorage> abvsStack;
+    protected final List<UTF8StringPointable> spStack;
+    protected final List<String> itemStack;
+
+    public JSONParser() throws JsonParseException, IOException {
+        factory = new JsonFactory();
+        atomic = new ArrayBackedValueStorage();
+        abStack = new ArrayList<ArrayBuilder>();
+        obStack = new ArrayList<ObjectBuilder>();
+        abvsStack = new ArrayList<ArrayBackedValueStorage>();
+        spStack = new ArrayList<UTF8StringPointable>();
+        itemStack = new ArrayList<String>();
+    }
+
+    public void parseDocument(File file, ArrayBackedValueStorage result)
+            throws NumberFormatException, JsonParseException, IOException {
+        DataOutput outResult = result.getDataOutput();
+        DataOutput out = atomic.getDataOutput();
+        StringValueBuilder svb = new StringValueBuilder();
+        String lastItem = null;
+        String startItem = null;
+        boolean nested = false;
+        JsonParser parser = factory.createParser(file);
+        while (!parser.isClosed()) {
+            JsonToken token = parser.nextToken();
+            if (token == JsonToken.START_ARRAY) {
+                if (!nested) {
+                    abvsStack.add(new ArrayBackedValueStorage());
+                    abStack.add(new ArrayBuilder());
+                }
+                itemStack.add("array");
+                lastItem = "array";
+                abvsStack.get(abvsStack.size() - 1).reset();
+                abStack.get(abStack.size() - 1).reset(abvsStack.get(abvsStack.size() - 1));
+            }
+            if (token == JsonToken.START_OBJECT) {
+                if (!nested) {
+                    abvsStack.add(new ArrayBackedValueStorage());
+                    obStack.add(new ObjectBuilder());
+                }
+                itemStack.add("object");
+                lastItem = "object";
+                abvsStack.get(abvsStack.size() - 1).reset();
+                obStack.get(obStack.size() - 1).reset(abvsStack.get(abvsStack.size() - 1));
+            }
+            if (token == JsonToken.FIELD_NAME) {
+                UTF8StringPointable sp = new UTF8StringPointable();
+                ArrayBackedValueStorage key = new ArrayBackedValueStorage();
```

Comment:
It seems that we should not need a new UTF8StringPointable and a new ArrayBackedValueStorage for each field name. Instead we should only need one per level of the depth of the JSON object. So that maximum number of UTF8StringPointables and ArrayBackedValueStorages that should be created is the nesting depth of JSON objects.

More generally, it seems useful for all objects that we're allocating to maintain a stack of these objects that only grows up to the maximal depth and that gets garbage collected at the end of the complete parsing process. To reuse those objects we just need to a) maintain the current depth that we're parsing at to find the object in the Stack and b) reset() the object before reusing it.